firefox_puppeteer/ui/tabbar.py

```diff
@@ -3,126 +3,315 @@
-                               .find_elements('tag name', 'tab'))
-        return [self.TabElement(tab) for tab in tabs]
+        # Get handles from tab exceptions, and find those which can be closed
+        for tab in self.tabs:
+            if tab not in exceptions:
+                tab.close(force=True)
+
+    def close_tab(self, tab=None, *args, **kwargs):
+        """Closes the tab by using the specified trigger.
 
-    def get_tab(self, target):
+        By default the currently selected tab will be closed. If another :class:`Tab`
+        is specified, that one will be closed instead. Also when the tab is closed a
+        :function:`switch_to` call is automatically performed, so that the new selected
+        tab becomes active.
+
+        :param tab: Optional, the :class:`Tab` instance to close. Defaults to
+         the currently selected tab.
+
+        :param trigger: Optional, method in how to close the current tab. This can
+         be a string with one of `menu` or `shortcut`, or a callback which gets triggered
+         with the :class:`Tab` as parameter. Defaults to `menu`.
+
+        :param force: Optional, forces the closing of the window by using the Gecko API.
+         Defaults to `False`.
         """
-        Get a reference to the specified tab.
+        tab = tab or self.selected_tab
+        tab.close(*args, **kwargs)
+
+    def open_tab(self, trigger='menu'):
+        """Opens a new tab in the current browser window.
+
+        If the tab opens in the foreground, a call to :function:`switch_to` will
+        automatically be performed. But if it opens in the background, the current
+        tab will keep its focus.
 
-        :param target: Either an index of `tabs` or a substring of the label.
-        :returns: A :class:`TabElement` corresponding to the specified tab.
+        :param trigger: Optional, method in how to open the new tab. This can
+         be a string with one of `menu`, `button` or `shortcut`, or a callback
+         which gets triggered with the current :class:`Tab` as parameter.
+         Defaults to `menu`.
+
+        :returns: :class:`Tab` instance for the opened tab
         """
-        if isinstance(target, int):
-            return self.tabs[target]
+        start_handles = self.marionette.window_handles
+
+        # Prepare action which triggers the opening of the browser window
+        if callable(trigger):
+            trigger(self.selected_tab)
+        elif trigger == 'button':
+            self.window.tabbar.newtab_button.click()
+        elif trigger == 'menu':
+            # TODO: Make use of menubar class once it supports ids
+            menu = self.window.marionette.find_element('id', 'menu_newNavigatorTab')
+            menu.click()
+        elif trigger == 'shortcut':
+            self.window.send_shortcut(self.window.get_localized_entity('tabCmd.commandkey'),
+                                      accel=True)
+        # elif - need to add other cases
```

Comment:
Yes, therefore we need more elements to work on, which will be part of the follow-up bug. Those are not that often used.